synapse/storage/databases/main/deviceinbox.py

```diff
@@ -664,21 +664,28 @@ def _remove_deleted_devices_from_device_inbox_txn(
             due to a single device having lots of device messages.
             """
 
-            last_stream_id = progress.get("stream_id", 0)
+            txn.execute("SELECT max(stream_id) FROM device_inbox")
+            row = txn.fetchone()
+            if row is None or row[0] is None:
+                max_stream_id = 0
+            else:
+                max_stream_id = row[0]
+
+            start = progress.get("stream_id", 0)
+            stop = start + batch_size
 
             sql = """
                 SELECT device_id, user_id, stream_id
                 FROM device_inbox
                 WHERE
-                    stream_id >= ?
+                    stream_id >= ? AND stream_id < ?
                     AND (device_id, user_id) NOT IN (
                         SELECT device_id, user_id FROM devices
                     )
                 ORDER BY stream_id
```

Comment:
I think the ORDER BY is redundant?

Reply:
In fact, why not just do a DELETE?

```
                DELETE FROM device_inbox
                WHERE
                    stream_id >= ? AND stream_id < ?
                    AND (device_id, user_id) NOT IN (
                        SELECT device_id, user_id FROM devices
                    )
```

Reply:
> I think the ORDER BY is redundant?

Ah so it is

> In fact, why not just do a DELETE?

We need to get data back to update the progress, but in a DELETE we can only do that using `RETURNING`, which SQLite < 3.35.0 (2021-03-12) doesn't support, and I don't think we want to break the update for them. So I'll split this with a condition on `supports_returning`, and do it in a single `DELETE` if the engine supports it (otherwise it'll do it in two steps like it does now).